### ISSUE TITLE
Change how trace and stream end times are calculated

### ISSUE THREAD

**Issue by SquirrelKnight** [obspy/obspy#3236]
### Problem

I have noticed that the end time for a stream or trace object is calculated by taking the start time and adding the sample rate * the number of samples. In most circumstances, this is acceptable, but there are situations where clock drift corrections have been applied to the miniseed data, and so the end time is incorrect.

### Proposed solution

I propose computing the endtime from the header in the last miniseed block of a file loaded into memory, rather than from the header in the first block. The same equation can be used as mentioned in the problem above (header time + (sample_rate*sample_num).

**Comment by paitor** [comment]
Hmmm,

would it not be more sensible (conservative) to convert each miniseed 
block to its own trace and the merge traces only if aligned to within 
set precision. The point being that if a drift correction have been 
applied from time t (i.e. a particular miniseed block) then obspy should 
not try to be smart and assume that this drift correction also applies 
to earlier blocks in the file.

**Comment by trichter** [comment]
If the clock drift is constant, you could also get the desired endtime by adjusting the sampling rate.
I feel calculating the endtime in a way proposed by @SquirrelKnight will break ObsPy in one way or another.

**Comment by d-chambers** [comment]
> I feel calculating the endtime in a way proposed by @SquirrelKnight will break ObsPy in one way or another.

I agree. But a parser-specific opt-in feature could be a reasonable compromise if someone has enough motivation to implement it. That way it won't change ObsPy's default behavior but can still address these issues when needed.

**Comment by SquirrelKnight** [comment]
Just for a bit more information, the drift correction is applied to each miniseed block. I imagine adjusting the sampling rate could cause havoc with filters, instrument response, and downsampling/upsampling.

**Comment by megies** [comment]
First of all, would be good to have a real world example showcasing the issue.

**Comment by megies** [comment]
> I have noticed that the end time for a stream or trace object is calculated by taking the start time and adding the the number of samples / the sample rate.

In a trace without gaps nothing else would make sense. It rather sounds like your issue might be deeper at that point where miniseed blocks get concatenated or not, and you're kind of expecting a gappy stream with a later endtime of the later traces or a stream with overlapping traces and an earlier endtime in one of the traces.

I don't know all details from the low level mseed reading, but there are some mechanisms to join blocks even when there technically are sub-sample gaps which could potentially sum up to a "more than one sample" gap in certain scenarios.

In any case, we should have an example showing the problem before discussing further, I think.

**Comment by SquirrelKnight** [comment]
I've zipped together a file with three subsequent daily mseed files that have been drift corrected. Each block has a drift corrected time. You can download the data from my Google Drive (the file is ~190 MB). Each miniseed file contains three component data from BN, HH, and MN channels, which run at different sampling rates of 25, 200, and 5 Hz, respectively.

https://drive.google.com/file/d/14EGDqLj3dciFdH_I-LsdYVrSDY2n94pp/view?usp=sharing

**Comment by SquirrelKnight** [comment]
Also, FYI, this is from an autonomous ocean-bottom seismometer. The deployment went from 2018-06-27T00:00:00 until 2019-09-20T22:00:00, during which time 4.42s of drift occurred.
